open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java

```diff
@@ -0,0 +1,214 @@
+    private static DeviceHandler deviceHandler;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getFriendlyName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest,
+                                                 HttpServletResponse httpServletResponse,
+                                                 AuthenticationContext authenticationContext)
+            throws AuthenticationFailedException {
+
+        authenticationContext.setCurrentAuthenticator(App2AppAuthenticatorConstants.AUTHENTICATOR_NAME);
+        String jwtString =
+                httpServletRequest.getParameter(App2AppAuthenticatorConstants.DEVICE_VERIFICATION_TOKEN_IDENTIFIER);
+
+        try {
+            SignedJWT signedJWT = JWTUtils.getSignedJWT(jwtString);
+            DeviceVerificationToken deviceVerificationToken = new DeviceVerificationToken(signedJWT);
+            //Extracting deviceId and loginHint is necessary to retrieve the public key
+            String loginHint = deviceVerificationToken.getLoginHint();
+            String deviceID = deviceVerificationToken.getDeviceId();
+
+            //Checking whether deviceId and loginHint present in passed jwt
+            if (StringUtils.isBlank(loginHint) || StringUtils.isBlank(deviceID)) {
+                throw new AuthenticationFailedException(App2AppAuthenticatorConstants.REQUIRED_PRAMAS_MISSING_MESSAGE);
+            }
+
+            AuthenticatedUser userToBeAuthenticated =
+                    App2AppAuthUtils.getAuthenticatedUserFromSubjectIdentifier(loginHint);
+            String publicKey = getPublicKeyByDeviceID(deviceID, userToBeAuthenticated);
+            deviceVerificationToken.setPublicKey(publicKey);
+            // setting the user is mandatory for data publishing purposes
+            //If exception is thrown before setting a user data publishing will encounter exceptions
+            authenticationContext.setSubject(userToBeAuthenticated);
+            /*
+                if validations are failed it will throw a JWTValidationException and flow will be interrupted.
+                Hence, user Authentication will fail.
+             */
+            App2AppAuthUtils.validateToken(deviceVerificationToken);
+            //If the flow is not interrupted user will be authenticated.
+            log.info(String.format(App2AppAuthenticatorConstants.USER_AUTHENTICATED_MSG,
+                    userToBeAuthenticated.getUserName()));
+        } catch (ParseException e) {
+            throw new AuthenticationFailedException
+                    (App2AppAuthenticatorConstants.PARSE_EXCEPTION_MESSAGE + e.getMessage(), e);
```

Comment:
Since we're already passing the exception, what is the usage of `e.getMessage()`?

Reply:
Resolved in [260e508](https://github.com/wso2/financial-open-banking/pull/67/commits/260e508da50ed60b02dc3fd2f2d27107b59691d2).